pin/gc/gc.go

```diff
@@ -0,0 +1,120 @@
+package gc
+
+import (
+	bstore "github.com/ipfs/go-ipfs/blocks/blockstore"
+	key "github.com/ipfs/go-ipfs/blocks/key"
+	bserv "github.com/ipfs/go-ipfs/blockservice"
+	offline "github.com/ipfs/go-ipfs/exchange/offline"
+	dag "github.com/ipfs/go-ipfs/merkledag"
+	pin "github.com/ipfs/go-ipfs/pin"
+
+	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
+	eventlog "github.com/ipfs/go-ipfs/thirdparty/eventlog"
+)
+
+var log = eventlog.Logger("gc")
+
+type GCSet struct {
+	keys map[key.Key]struct{}
+}
+
+func NewGCSet() *GCSet {
+	return &GCSet{make(map[key.Key]struct{})}
+}
+
+func (gcs *GCSet) Add(k key.Key) {
+	gcs.keys[k] = struct{}{}
+}
+
+func (gcs *GCSet) Has(k key.Key) bool {
+	_, has := gcs.keys[k]
+	return has
+}
+
+func (gcs *GCSet) AddDag(ds dag.DAGService, root key.Key) error {
+	ctx := context.Background()
+	nd, err := ds.Get(ctx, root)
+	if err != nil {
+		return err
+	}
+
+	gcs.Add(root)
+
+	for _, lnk := range nd.Links {
+		k := key.Key(lnk.Hash)
+		err := gcs.AddDag(ds, k)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
my impression is that without concurrency, this will trade off between IO + memory and lose time in the context switches with the kernel. Perhaps one good hack might be to have 2+ concurrent goroutines calling AddDag? would need some mutex on the map, so that might actually be enough to kill the improvement... but it _is_ IO + _lots_ of syscalls that we might schedule better...